### DISCUSSION TITLE
Depth Supervision

### DISCUSSION THREAD

**Discussion by reynoldscem** [NVlabs/instant-ngp#647] [Q&A]
I saw in the recent commits that depth supervision was added. Could you please inform me how to go about using this or provide an example?

**Comment by Karbo123** [comment]
I have the same question, and I don't know whether the function of depth supervision have been finished. I have tried several time, and I just found that if we combine RGB image supervision with Depth supervision, the visualzation results will be even worse. My depth images are rendered by enabling `testbed.render_mode = ngp.RenderMode.Depth`, and saved as `.npy` format. I'm not sure if something is wrong for my depth maps. I used `depth_scale=1`.

**Reply by Karbo123** [reply]
Okey, I just found that, the depth map should be normalized as well, I used `depth_scale=0.33` for the LEGO dataset, because `testbed.nerf.training.dataset.scale == 0.33000001311302185`. After normalizing the depth maps, everything just works fine.

**Reply by Karbo123** [reply]
And I also found learning from depth supervision do help fitting. I trained on 10 images, 1600 iterations (fast training, only 10sec), and results are as follows.
With Depth: PSNR = 26.981 (min=21.356;max=33.327), MSE = 2.527e-03, SSIM = 8.05722e-01
Without Depth: PSNR = 24.771 (min=20.470;max=32.369), MSE = 3.876e-03, SSIM = 8.14602e-01

**Reply by Karbo123** [reply]
I finally found that the magic number `0.33` comes from `static constexpr float NERF_SCALE = 0.33f;`, it always seems to be `0.33`, regardless of the dataset we actually used.

**Reply by mbennehar** [reply]
Hi, 
Thanks for your comments, they are very helpful. Do you mind sharing the LEGO dataset with the depth that you generated for training with depth support? Also, what value of lambda did you use to balance the RGB and depth losses?
Thanks.

**Reply by Karbo123** [reply]
@mrmouss No sorry, but you can write the depth generation script on your own, it's pretty simple. All you need to do is just to fit the LEGO dataset with enough number of iterations, and then render the depth maps on the training set by using `testbed.render_mode = ngp.RenderMode.Depth`, the output images are alreay the real depth values, just simply pick one of the channels because I found the first three channels seem to be the same (e.g. `depth[..., 0]`), and I'd suggest to save them in `.npy` format. Although the depth maps are not very accurate, but I think they're enough for most cases. BTW, I simply use `lambda=1.0`, and obviously, it's chosen without any careful consideration.

**Reply by mbennehar** [reply]
Hi,
I just noticed that the method that loads the depth already accounts for the dataset scale:
https://github.com/NVlabs/instant-ngp/blob/07d8e2ca7232f97397ab73af9b56c7db639d3445/src/nerf_loader.cu#L699
		result.set_training_image(i, m.res, m.pixels, m.depth_pixels, **m.depth_scale * result.scale**, m.image_data_on_gpu, m.image_type, EDepthDataType::UShort, sharpen_amount, m.white_transparent, m.black_transparent, m.mask_color, m.rays);

**Reply by Karbo123** [reply]
@mrmouss well, then I think maybe you should just set depth scale to an appropriate value by giving `integer_depth_scale`. however, in my case, I call `Testbed::Nerf::Training::set_image` to set images, and the depth values are already the actual real ones, so I just simply set it to 0.33.

**Reply by mbennehar** [reply]
Makes sense, thanks for your reply.

**Reply by roangel** [reply]
Do I get this correctly that what you are doing is:
1. Train a normal nerf with normal pictures
2. Get the depth map from the trained nerf
3. Keep training the nerf using these nerf maps

I am wondering, what if I do have the depth groundtruth (input images are RGBD), can it be used directly for the nerf generation? Would this significantly improve the depth map that is obtained from the nerf?
If so, can you provide an example? Thanks!

**Reply by Karbo123** [reply]
@roangel Hi, if you already have GT depth maps, you can just directly use them by fitting the depth values as well. I think the quality will be improved if the provided depth supervision is accurate. But if the depth images are noisy, it's hard to say, the qualitative results could probably drop.

**Reply by JasonLSC** [reply]
Hi, if I already have GT depth maps, do I need to set `depth_scale=1`  or `integer_depth_scale=1`?

**Reply by Karbo123** [reply]
@JasonLSC If you have depth images already, and if you put the depth images appropriately under a folder, and load these by providing the dataset config json file, you should probably set `integer_depth_scale`; but if you load them by calling `Testbed::Nerf::Training::set_image` and loading them from numpy array, then you should set `depth_scale`. I would highly encourage you to check the source codes, and figure out by yourself how it loads depth images into the CUDA/C++ backend.

**Reply by CherniGhassen** [reply]
@Karbo123  Hi, to add the depth supervision, you create a second json file specific to the depth data? And what command to you write to use them? I'm having trouble understanding how to use the depth and I'm not very familiar with .cu files

**Reply by AIBluefisher** [reply]
> @JasonLSC If you have depth images already, and if you put the depth images appropriately under a folder, and load these by providing the dataset config json file, you should probably set `integer_depth_scale`; but if you load them by calling `Testbed::Nerf::Training::set_image` and loading them from numpy array, then you should set `depth_scale`. I would highly encourage you to check the source codes, and figure out by yourself how it loads depth images into the CUDA/C++ backend.

If I have depth values from colmap, then do I need to normalize the depth map and add the depth_scale? I think the depth_scale should be tuned to match the sampling interval of depths, but I have no idea how the sampling interval of instant_ngp is determined.

**Reply by SSground** [reply]
> > @JasonLSC If you have depth images already, and if you put the depth images appropriately under a folder, and load these by providing the dataset config json file, you should probably set `integer_depth_scale`; but if you load them by calling `Testbed::Nerf::Training::set_image` and loading them from numpy array, then you should set `depth_scale`. I would highly encourage you to check the source codes, and figure out by yourself how it loads depth images into the CUDA/C++ backend.
> 
> If I have depth values from colmap, then do I need to normalize the depth map and add the depth_scale? I think the depth_scale should be tuned to match the sampling interval of depths, but I have no idea how the sampling interval of instant_ngp is determined.

Can you tell me how to get the depth values ​​that colmap finds. Are there any relevant script.

**Reply by chl2** [reply]
Does the depth information you are talking about mean that each image must have a depth image? Can I use colmap's 3d point, which is the overall depth information?

**Reply by Karbo123** [reply]
@chl2 Yes, for sparse depth supervision, maybe you can take a look at [DSNeRF](https://github.com/dunbar12138/DSNeRF) which uses sparse point cloud from COLMAP. For instant-ngp, you should provide depth images (you can render the point cloud into multi-view depth images).

**Reply by ddkats** [reply]
Hi guys,
When I enable depth supervision, the result seems to be worse (visualize black spots and blurry outcome). I think that this has to do to with the format of the depth images, I mean as far as I can understand ingp expects a float 16bit format in pixels, normalized at [0,1]. Is that correct? Can anyone provide an example depth map to see the format used?  @Karbo123 

(*the depth maps I use are generated by a stereo camera and I have corresponding depth and rgb images)

**Reply by kafei123456** [reply]
![1697678769189](https://github.com/NVlabs/instant-ngp/assets/73866655/bf9e16da-bad5-4fda-87c4-472fb02b20bf)
I set the transforms.json like that , but it have not loaded depth map. How to load depth map ?

**Comment by CherniGhassen** [comment]
Did you manage to add depth supervision? There's no explanation anywhere  on how to add it

**Reply by Karbo123** [reply]
Hello, if you are familiar with c++, I'd recommend you read the source code yourself at [here](https://github.com/NVlabs/instant-ngp/blob/master/src/nerf_loader.cu).

Especially have a look at (load from disk where files are specified by a `json` file)

https://github.com/NVlabs/instant-ngp/blob/107669abe0d3a270a07e79426a616e767caf6631/src/nerf_loader.cu#L426-L429

https://github.com/NVlabs/instant-ngp/blob/107669abe0d3a270a07e79426a616e767caf6631/src/nerf_loader.cu#L477-L479

https://github.com/NVlabs/instant-ngp/blob/107669abe0d3a270a07e79426a616e767caf6631/src/nerf_loader.cu#L616-L629

or even here if you like to dig more about it:

https://github.com/NVlabs/instant-ngp/blob/107669abe0d3a270a07e79426a616e767caf6631/src/nerf_loader.cu#L735-L835

**Reply by CherniGhassen** [reply]
Thank you very much, this is helpful

**Reply by poyodiaz** [reply]
> Did you manage to add depth supervision? There's no explanation anywhere on how to add it

Hi i added  : 

"enable_depth_loading": true,
 "integer_depth_scale": 1,
"depth_path": "./depth/0002-standard-scale-4_00x.png",

to my transforms json file but it doesnt load at all could you please help me to understand what am i missing?

**Reply by ddkats** [reply]
> > Did you manage to add depth supervision? There's no explanation anywhere on how to add it
> 
> Hi i added :
> 
> "enable_depth_loading": true, "integer_depth_scale": 1, "depth_path": "./depth/0002-standard-scale-4_00x.png",
> 
> to my transforms json file but it doesnt load at all could you please help me to understand what am i missing?

You need to have corresponding depth map image for each input image in the transform.json. Then, when you open the gui (or headless) I think it's necessary to adjust depth_supervision value. 
![Screenshot_14](https://github.com/NVlabs/instant-ngp/assets/38035561/93608c3a-16ac-4d3f-b328-2c1786941b08)

**Reply by poyodiaz** [reply]
ok i see , thanks ! so the json must contain the info on each frame along the whole file ?

**Reply by ddkats** [reply]
I think so...Also I'm waiting for a reply at this thread here:
https://github.com/NVlabs/instant-ngp/issues/1095
